README.md
# bedrock-rest

A [bedrock][] module that adds helpers for REST support.

## Requirements

- npm v3+

## Quick Examples

```
npm install bedrock-rest
```

```js
var bedrock = require('bedrock');
var brIdentity = require('bedrock-identity');
var brPassport = require('bedrock-passport');
var brRest = require('bedrock-rest');
var cors = require('cors');
var myItems = require('my-items-library');

bedrock.events.on('bedrock-express.init', function(app) {
  app.param('identity', brRest.idParam);
});

bedrock.events.on('bedrock-express.configure.routes', addRoutes);

function addRoutes(app) {
  var idPath = bedrock.config.identity.basePath + '/:identity';

  app.get(idPath + '/dashboard',
    brPassport.ensureAuthenticated,
    brRest.makeResourceHandler());

  app.options(idPath + '/items', cors());
  app.get(idPath + '/items',
    cors(),
    brRest.makeResourceHandler({
      get: function(req, res, callback) {
        var identityId = brIdentity.createIdentityId(req.params.identity);

        myItems.getItems(identityId, function(err, items) {
          if(err) {
            return callback(err);
          }
          callback(null, items);
        });
      }
    }));

  app.options(idPath + '/items/:item', cors());
  app.get(idPath + '/items/:item',
    cors(),
    brRest.makeResourceHandler({
      get: function(req, res, callback) {
        var identityId = brIdentity.createIdentityId(req.params.identity);
        var itemId = myItems.createItemId(identityId, req.params.item);

        myItems.getItem({id: itemId}, function(err, item) {
          callback(err, item);
        });
      }
    }));
}
```

## API

### idParam(req, res, next, id)

Validates an ID from a URL path and, it passes validation, it will be
available via req.params. This method is for use with [express][] or
[bedrock-express][]:

```js
// passed to an express server's param call
server.param(':foo', rest.idParam);

// setup with "bedrock-express.init" event
bedrock.events.on('bedrock-express.init', function(app) {
  app.param('foo', rest.idParam);
});
```

### makeResourceHandler(options)

Make middleware for a type negotiated REST resource. This middleware handles
the details of handling requests for `json`, `application/ld+json`, and
`html`. JSON based requests will just return the data from the `get` option.
HTML requests default to using another route to return HTML. It defaults to
`main.html` which could be setup to start a single page app that calls the
same endpoint to get the JSON based resource.

`options`:
- **validate**: content to pass to bedrock.validation.validate
- **get(req, res, callback(err, data))**: get resource data
- **json**: handle JSON content types (false, true, 'route') (default: true)
- **html**: handle HTML content types (false, 'route') (default: 'route')

```js
makeResourceHandler({
  get: function(req, res, callback) {
    myLib.load(..., function(err, resource) {
      if(err) {
        return callback(err);
      }
      callback(null, resource);
    }
  }
});
```

### linkedDataHandler(options)

Alias for makeResourceHandler.

### when.prefers.jsonld

Middleware to check if request prefers JSON-LD

### when.prefers.ld

Middleware to check if request prefers linked data.

[bedrock]: https://github.com/digitalbazaar/bedrock
[bedrock-express]: https://github.com/digitalbazaar/bedrock-express
[express]: https://github.com/strongloop/express


CHANGELOG.md
# bedrock-rest ChangeLog

## 3.0.0 - 2019-11-08

### Added
- eslint support.

### Changed
- **BREAKING**: `makeResourceHandler` options changed.
  - No longer takes template related options in favor of routing to next
    middleware.
  - Added `json` option to turn on/off JSON processing or route to next
    middleware.
  - Added `html` option to turn off HTML processing or route to next
    middleware.
- Code modernization and improvements.
- Reuse `makeResourceHandler` for `linkedDataHandler` implementation.

### Removed
- `bedrock-views` dependency due to no longer needing view vars feature.

## 2.1.3 - 2019-03-26

### Changed
- Update bedrock-validation peer dependency.

## 2.1.2 - 2018-09-20

### Fixed
- Fix typo in peer dependencies.

## 2.1.1 - 2018-09-20

### Changed
- Update bedrock-validation peer dependency.

## 2.1.0 - 2017-06-27

### Changed
- Update `bedrock-views` dependency from 4.x to 5.x.

## 2.0.2 - 2016-09-28

### Fixed
- If `application/ld+json`, `application/json` are not preferred
  over `text/html` return false from `prefers`.

## 2.0.1 - 2016-03-15

### Changed
- Update bedrock dependencies.

## 2.0.0 - 2016-03-03

### Changed
- Update package dependencies for npm v3 compatibility.

## 1.1.0 - 2015-11-16

### Added
- Add `when` middleware.

## 1.0.1 - 2015-05-07

## 1.0.0 - 2015-04-08

## 0.1.0 (up to early 2015)

- See git history for changes.


lib/index.js
/*
 * Bedrock REST module.
 *
 * Copyright (c) 2012-2018 Digital Bazaar, Inc. All rights reserved.
 */
const bedrock = require('bedrock');
const _validate = require('bedrock-validation').validate;
const BedrockError = bedrock.util.BedrockError;

// module API
const api = {};
module.exports = api;

/**
 * List of default acceptable "Accept" header MIME types for JSON-LD.
 */
const DEFAULT_JSON_LD_TYPES = [
  'application/ld+json',
  'application/json'
];

/**
 * List of default acceptable "Accept" header MIME types for linked data.
 */
const DEFAULT_LINKED_DATA_TYPES = [
  'application/ld+json',
  'application/json'
  // TODO: support other types
];

/**
 * List of default acceptable "Accept" header MIME types for HTML.
 */
const DEFAULT_HTML_TYPES = [
  'text/html'
];

/**
 * List of default acceptable "Accept" header MIME types.
 */
const DEFAULT_ACCEPTABLE_TYPES = [
  ...DEFAULT_HTML_TYPES,
  ...DEFAULT_LINKED_DATA_TYPES
];

/**
 * Validates an ID from a URL path and, it passes validation, it will
 * be available via req.params. This method should be passed to an express
 * server's param call, eg:
 *
 * server.param(':foo', idParam)
 *
 * @param req the request.
 * @parma res the response.
 * @param next the next handler.
 * @param id the id.
 */
api.idParam = function(req, res, next, id) {
  const regex = /[a-zA-Z0-9][\-a-zA-Z0-9~_\.]*/;
  if(!regex.test(id)) {
    res.redirect('/');
  } else {
    next();
  }
};

/**
 * Make middleware for a type negotiated REST resource.
 *
 * FIXME: Add docs.
 *
 * JSON handling can be controlled with the 'json' option. A boolean value will
 * turn on or off processing. The string 'route' will use the next middleware.
 *
 * HTML handling can be controlled with the 'html' option. False will turn off
 * processing. The string 'route' will use the next middleware.
 *
 * @param options the handler options.
 *   validate: content to pass to bedrock.validation.validate
 *   get(req, res, callback(err, data)): get resource data
 *   json: handle JSON content types (false, true, 'route') (default: true)
 *   html: handle HTML content types (false, 'route') (default: 'route')
 */
api.makeResourceHandler = function({
  validate = null,
  get,
  json = true,
  html = 'route'
}) {
  const middleware = [];
  // optional validation
  if(validate) {
    middleware.push(_validate(validate));
  }
  // main handler
  middleware.push(function(req, res, next) {
    function _json() {
      if(!get) {
        res.sendStatus(204);
        return;
      }
      get(req, res, function(err, data) {
        if(err) {
          next(err);
          return;
        }
        res.json(data);
      });
    }
    function _route() {
      next('route');
    }

    const formatOptions = {};
    const acceptable = [];
    if(json) {
      DEFAULT_JSON_LD_TYPES.forEach(type => {
        formatOptions[type] = json === 'route' ? _route : _json;
        acceptable.push(type);
      });
    }
    if(html === 'route') {
      DEFAULT_HTML_TYPES.forEach(type => {
        formatOptions[type] = _route;
        acceptable.push(type);
      });
    }
    formatOptions.default = function() {
      next(new BedrockError(
        'Requested content types not acceptable.',
        'NotAcceptable', {
          public: true,
          httpStatusCode: 406,
          details: {
            acceptable
          }
        }));
    };
    res.format(formatOptions);
  });
  return middleware;
};

/**
 * Compatilibity alias for makeResourceHandler.
 */
api.linkedDataHandler = api.makeResourceHandler;

/**
 * Creates a middleware that will call a boolean function to check if a
 * request is acceptable for subsequently attached middlewares (middlewares
 * that are attached using the same call that attaches this middleware). If it
 * is, next() is called, if not, next('route') is called.
 *
 * This function can also be used as a collection of when functions.
 *   when.prefers.json
 *
 * @param check(req, res): function to check a request. Returns boolean.
 */
api.when = function(check) {
  return function(req, res, next) {
    if(check(req, res)) {
      return next();
    }
    next('route');
  };
};

/**
 * Adds a `prefers` shortcut to `when` to check if, given a set of acceptable
 * types, if a request prefers a type from a list.
 *
 * @param acceptable the known acceptable types for a route.
 * @param preferred a list of preferred types.
 */
api.when.prefers = function(acceptable, preferred) {
  // TODO: remove "acceptable" parameter if not needed
  // TODO: support spread parameters for preferred
  if(!Array.isArray(preferred)) {
    preferred = [preferred];
  }
  /* eslint-disable-next-line no-unused-vars */
  return api.when(function(req, res) {
    return preferred.indexOf(req.accepts(acceptable)) !== -1;
  });
};

/**
 * Adds a `jsonld` shortcut to `when.prefers` to check if a request prefers
 * JSON-LD.
 */
api.when.prefers.jsonld = api.when.prefers(
  DEFAULT_ACCEPTABLE_TYPES, DEFAULT_JSON_LD_TYPES);

/**
 * Adds a `ld` shortcut to `when.prefers` to check if a request prefers
 * linked data.
 */
api.when.prefers.ld = api.when.prefers(
  DEFAULT_ACCEPTABLE_TYPES, DEFAULT_LINKED_DATA_TYPES);


lib/rest.js
/*
 * Bedrock REST module.
 *
 * Copyright (c) 2012-2015 Digital Bazaar, Inc. All rights reserved.
 */
var async = require('async');
var bedrock = require('bedrock');
var validate = require('bedrock-validation').validate;
var views = require('bedrock-views');
var BedrockError = bedrock.util.BedrockError;

// module API
var api = {};
module.exports = api;

/**
 * List of default acceptable "Accept" header MIME types for linked data.
 */
var DEFAULT_LINKED_DATA_TYPES = [
  'application/ld+json',
  'application/json'
];

/**
 * List of default acceptable "Accept" header MIME types.
 */
var DEFAULT_ACCEPTABLE_TYPES = [
  'application/ld+json',
  'application/json',
  'text/html'
];

/**
 * Validates an ID from a URL path and, it passes validation, it will
 * be available via req.params. This method should be passed to an express
 * server's param call, eg:
 *
 * server.param(':foo', idParam)
 *
 * @param req the request.
 * @parma res the response.
 * @param next the next handler.
 * @param id the id.
 */
api.idParam = function(req, res, next, id) {
  var regex = /[a-zA-Z0-9][\-a-zA-Z0-9~_\.]*/;
  if(!regex.test(id)) {
    res.redirect('/');
  } else {
    next();
  }
};

/**
 * Make middleware for a type negotiated REST resource.
 *
 * FIXME: Add docs.
 *
 * @param options the handler options.
 *   validate: content to pass to bedrock.validation.validate
 *   get(req, res, callback(err, data)): get resource data
 *   template: template file name for HTML mode (default: 'main.html')
 *   templateNeedsResource: boolean flag to get resource for template
 *   updateVars(resource, vars, callback(err)): update vars with resource
 */
api.makeResourceHandler = function(options) {
  options = options || {};
  var middleware = [];
  // optional validation
  if(options.validate) {
    middleware.push(validate(options.validate));
  }
  // main handler
  middleware.push(function(req, res, next) {
    function _json() {
      if(!options.get) {
        res.sendStatus(204);
        return;
      }
      _getResource(req, res, options, function(err, data) {
        if(err) {
          return next(err);
        }
        res.json(data);
      });
    }
    function _html() {
      async.auto({
        vars: function(callback) {
          views.getDefaultViewVars(req, callback);
        },
        resource: ['vars', function(callback) {
          if(options.templateNeedsResource) {
            return _getResource(req, res, options, callback);
          }
          callback();
        }],
        updateVars: ['resource', function(callback, results) {
          if(options.templateNeedsResource && options.updateVars) {
            return options.updateVars(
              results.resource, results.vars, callback);
          }
          callback();
        }]
      }, function(err, results) {
        if(err) {
          return next(err);
        }
        // FIXME use option for template name
        res.render(options.template || 'main.html', results.vars);
      });
    }
    res.format({
      'application/ld+json': _json,
      json: _json,
      html: _html,
      'default': function() {
        next(new BedrockError(
          'Requested content types not acceptable.',
          'NotAcceptable', {
            'public': true,
            httpStatusCode: 406,
            details: {
              acceptable: [
                'application/json',
                'application/ld+json',
                'text/html'
              ]
            }
          }));
      }
    });
  });
  return async.applyEachSeries(middleware);
};

/**
 * Wrapper to get resources and handle error mapping.
 */
function _getResource(req, res, options, callback) {
  return options.get(req, res, function(err, data) {
    if(err) {
      return callback(err);
    }
    callback(null, data);
  });
}

/**
 * Creates a middleware that will call a boolean function to check if a
 * request is acceptable for subsequently attached middlewares (middlewares
 * that are attached using the same call that attaches this middleware). If it
 * is, next() is called, if not, next('route') is called.
 *
 * This function can also be used as a collection of when functions.
 *   when.prefers.json
 *
 * @param check(req, res): function to check a request. Returns boolean.
 */
api.when = function(check) {
  return function(req, res, next) {
    if(check(req, res)) {
      return next();
    }
    next('route');
  };
};

/**
 * Adds a `prefers` shortcut to `when` to check if, given a set of acceptable
 * types, if a request prefers a type from a list.
 *
 * @param acceptable the known acceptable types for a route.
 * @param preferred a list of preferred types.
 */
api.when.prefers = function(acceptable, preferred) {
  // TODO: remove "acceptable" parameter if not needed
  // TODO: support spread parameters for preferred
  if(!Array.isArray(preferred)) {
    preferred = [preferred];
  }
  return api.when(function(req, res) {
    return preferred.indexOf(req.accepts(acceptable)) !== -1;
  });
};

/**
 * Adds an `ld` shortcut to `when.prefers` to check if a request prefers
 * json-ld.
 */
api.when.prefers.jsonld = api.when.prefers(
  DEFAULT_ACCEPTABLE_TYPES, ['application/ld+json', 'application/json']);

/**
 * Adds an `ld` shortcut to `when.prefers` to check if a request prefers
 * linked data.
 */
api.when.prefers.ld = api.when.prefers(
  DEFAULT_ACCEPTABLE_TYPES, DEFAULT_LINKED_DATA_TYPES);

/**
 * Make middleware for a type negotiated REST resource.
 *
 * FIXME: Add docs.
 *
 * @param options the handler options.
 *   validate: content to pass to bedrock.validation.validate
 *   get(req, res, callback(err, data)): get resource data
 */
api.linkedDataHandler = function(options) {
  options = options || {};
  var middleware = [];
  // optional validation
  if(options.validate) {
    middleware.push(validate(options.validate));
  }
  // main handler
  middleware.push(function(req, res, next) {
    function _json() {
      if(!options.get) {
        res.sendStatus(204);
        return;
      }
      _getResource(req, res, options, function(err, data) {
        if(err) {
          return next(err);
        }
        res.json(data);
      });
    }
    res.format({
      'application/ld+json': _json,
      json: _json,
      'default': function() {
        next(new BedrockError(
          'Requested content types not acceptable.',
          'NotAcceptable', {
            'public': true,
            httpStatusCode: 406,
            details: {
              acceptable: DEFAULT_ACCEPTABLE_TYPES
            }
          }));
      }
    });
  });
  return async.applyEachSeries(middleware);
};
